### PR TITLE
#8454 Fix for std.mem.replacementSize adjacent matches bug.

### PR DESCRIPTION
**Pull request by jumpnbrownweasel**
When two 'needle' values are adjacent in the 'input' slice, the size is not
counted correctly. The 2nd 'needle' value is not matched because the index is
incremented by one after changing the index to account for the first value.

The impact is that the size returned is incorrect, which could cause UB when this
size is used as the size of the output buffer passed to std.mem.replace.

### CODE REVIEW

File: lib/std/mem.zig [hunk 1]
```diff
@@ -1899,22 +1899,38 @@ pub fn replace(comptime T: type, input: []const T, needle: []const T, replacemen
...
 /// Calculate the size needed in an output buffer to perform a replacement.
 pub fn replacementSize(comptime T: type, input: []const T, needle: []const T, replacement: []const T) usize {
     var i: usize = 0;
     var size: usize = input.len;
-    while (i < input.len) : (i += 1) {
+    while (i < input.len) {
         if (mem.indexOf(T, input[i..], needle) == @as(usize, 0)) {
             size = size - needle.len + replacement.len;
             i += needle.len;
+        } else {
```
Comment (LemonBoy) [inline lib/std/mem.zig:1932]: `std.mem.replacementSize(u8, "foo", "", "a")` will now enter an endless loop.
Passing an empty string as needle is somewhat useful to intersperse the replacement between each item, we may want to handle it or filter it out with an assertion.

File: lib/std/mem.zig [hunk 2]
```diff
@@ -1899,22 +1899,38 @@ pub fn replace(comptime T: type, input: []const T, needle: []const T, replacemen
...
+
+    replacements = replace(u8, "\\n\\n", "\\n", "\n", output[0..]);
+    expected = "\n\n";
+    testing.expect(replacements == 2);
+    testing.expect(eql(u8, output[0..expected.len], expected));
+
+    replacements = replace(u8, "abbba", "b", "cd", output[0..]);
+    expected = "acdcdcda";
+    testing.expect(replacements == 3);
+    testing.expect(eql(u8, output[0..expected.len], expected));
```
Comment (LemonBoy) [inline lib/std/mem.zig:1921]: Nit, use expectEqualStrings in place of expect + eql.

File: lib/std/mem.zig (hunk 1 above)
Comment (cryptocode) [reply lib/std/mem.zig:1932]: Rust, Go, JS, etc support this, and I think Zig should too. It currently doesn't on master (at least `replaceOwned` panics if given an empty needle), so maybe that could be a follow-up PR and just assert in this one.

Comment (jumpnbrownweasel) [reply lib/std/mem.zig:1932]: I prefer to add an assert for this for now, and implement "intersperse" as a separate PR. I tried this (empty needle) with mem.replace and it just loops until the output buffer overflows. So we currently don't support this in any form. I'll add the assert to all the appropriate places and retest.

File: lib/std/mem.zig (hunk 2 above)
Comment (jumpnbrownweasel) [reply lib/std/mem.zig:1921]: Will do.

File: lib/std/mem.zig (hunk 1 above)
Comment (cryptocode) [reply lib/std/mem.zig:1932]: I think that plan makes sense 👍

**Review by cryptocode — APPROVED** [review]
LGTM, tested locally and it solves the original issue I was having

**Comment by jumpnbrownweasel** [comment]
@LemonBoy Is there anything else you need from me for review? @cryptocode has approved it.

**Comment by andrewrk** [comment]
Thanks @jumpnbrownweasel!

**Pull request merged**